pkg/certificate/certificate.go

```diff
@@ -1,206 +1,93 @@
-	}
-
-	caPrivateKey, err := LoadPrivateKey(filepath.Join(caPath, "rootCA-key.pem"))
-	if err != nil {
-		return nil, wrapAndPrint("while loading CA key", err)
-	}
-
-	certBytes, privateKey, err := GenerateSignedCertificate(hello.ServerName, privateKey, caCertificate, caPrivateKey)
-	if err != nil {
-		return nil, wrapAndPrint("while generating signed certificate", err)
-	}
-
-	var cert tls.Certificate
-
-	cert.Certificate = append(cert.Certificate, certBytes)
-	cert.PrivateKey = privateKey
-
-	return &cert, nil
-}
-
-// GenerateSignedCertificate generates a certificate and signed it with the given AC.
-func GenerateSignedCertificate(
-	serverName string,
-	privateKey *rsa.PrivateKey,
-	caCertificate *x509.Certificate, caPrivateKey crypto.PrivateKey,
-) ([]byte, *rsa.PrivateKey, error) {
-	if len(serverName) == 0 {
-		return nil, nil, errors.New("while generating certificate: server name is empty")
-	}
+var (
 
-	now, err := time.Now().MarshalBinary()
-	if err != nil {
-		return nil, nil, fmt.Errorf("while marshaling now: %w", err)
-	}
-
-	uniqueSiteNameID := sha256.New().Sum(append([]byte(serverName), now...))
+	// ErrFailedToReadFile is returned when the provided file cannot be read.
+	ErrFailedToReadFile = errors.New("unable to read file")
+	// ErrFailedToDecodeFile is returned when the provided file cannot be decoded.
+	ErrFailedToDecodeFile = errors.New("unable to decode file content")
+	// ErrFailedToParseContent is returned when the provided content cannot be parsed.
+	ErrFailedToParseContent = errors.New("unable to parse content")
+)
 
-	//nolint:exhaustruct
-	template := &x509.Certificate{
-		//nolint:lll
-		SerialNumber: new(big.Int).SetBytes(uniqueSiteNameID), // in order to have an unique ID for each website Name, we hash the website name
-		Subject: pkix.Name{ // not necessary, but cool to have if the user ask for certificate details.
-			CommonName:   serverName,
-			Organization: []string{"station dynamically generated"},
-		},
-		NotBefore: time.Now(),
-		// one day of validity is enough since the certificate is generated dynamically each time
-		NotAfter: time.Now().AddDate(0, 0, 1),
+// readFile reads the content of a file at the given path.
+func readFile(path string) ([]byte, error) {
+	if path == "" {
+		return nil, fmt.Errorf("%w: %s", ErrFailedToReadFile, "empty path")
 	}
 
-	template.DNSNames = append(template.DNSNames, serverName)
-
-	// Create the certificate.
-	cert, err := x509.CreateCertificate(rand.Reader, template, caCertificate, privateKey.Public(), caPrivateKey)
+	raw, err := os.ReadFile(path)
 	if err != nil {
-		return nil, nil, fmt.Errorf("while creating certificate: %w", err)
-	}
-
-	return cert, privateKey, nil
-}
-
-// Get the mkcert CA root path depending on the OS.
-func mkcertCARootPath() (string, error) {
-	if env := os.Getenv("CAROOT"); env != "" {
-		return env, nil
+		return nil, fmt.Errorf("%w at %s: %w", ErrFailedToReadFile, path, err)
```

Comment:
Looks like you can't have multiple occurences of `%w` in a `fmt.Errorf` and that's why tests are failing

Reply:
> Looks like you can't have multiple occurences of `%w` in a `fmt.Errorf` and that's why tests are failing

This functionality was added in go 1.20.

Why do i need both ? To let errors.As be usable.